python/cog/json.py

```diff
@@ -61,9 +61,9 @@ def upload_files(obj: Any, upload_file: Callable[[io.IOBase], str]) -> Any:
         return {key: upload_files(value, upload_file) for key, value in obj.items()}
     if isinstance(obj, list):
         return [upload_files(value, upload_file) for value in obj]
-    if isinstance(obj, Path):
-        with obj.open("rb") as f:
-            return upload_file(f)
     if isinstance(obj, io.IOBase):
         return upload_file(obj)
+    if callable(getattr(obj, "open", None)):
+        with obj.open("rb") as f:
+            return upload_file(f)
```

Comment:
This makes me a bit nervous. The contract isn't purely that the object conforms to `open()` right? It also needs `seek`, `read` etc...

Reply:
@aron it's definitely a bit more loose than I'd like 😅 I'm happy to tighten it up.

Reply:
Hmm, I think it's okay, perhaps we should assert that `f` is an instance of `io.IOBase`…

Reply:
mmm yeah I like that

Reply:
Asking coz cog-runtime doesn't handle upload/download in Python, but rather in the Go server. So the `Path` seen by Python predictor code is always a local file. The Python runner serializes it as `file:///<path>` in the JSON output file & passes it to Go server, which then picks up the file for upload.
Just for context so we keep it in mind for converging cog runtimes later.

Reply:
@meatballhat I think we're looking at this wrong. `cog.Path` really only exists to provide a Pydantic variant of `pathlib.Path` for schema generation/validation (If we ignore the `URLPath` shenanigans).

Anywhere we are dealing with outputs I think we only care about `pathlib.Path`, so that's probably the simplest change here.

Reply:
@aron I'm good with assuming `pathlib.Path`, but in order to handle both that and `cog.types.Path` we'll need to do something like this change to move away from `isinstance`, or use `isinstance` with a typing protocol that both `pathlib.Path` and `cog.types.Path` fulfill.

Reply:
Partial retraction:

```
[ins] In [6]: cog_path = cog.types.Path("./ok-bud")

[ins] In [7]: pathlib_path = pathlib.Path("./ok-bud")

[ins] In [8]: isinstance(cog_path, pathlib.Path)
Out[8]: True

[ins] In [9]: isinstance(pathlib_path, cog.types.Path)
Out[9]: False 

```

Reply:
Yeah, this is what I meant. I think as long as we don't extend `cog.Path` with behaviour beyond the Pydantic types we're fine in this specific case and I think pretty much in the general case too.

Reply:
I think my extreme dislike of `isinstance` is showing 😅

Reply:
You nerd-sniped me. It looks like Python 3.6 introduced `os.PathLike` so you can detect a path-like object (that can be passed to `open()`) by implementing `__fspath__()`. Though I still think that probably still requires an `isinstance` check...

https://docs.python.org/3/whatsnew/3.6.html#pep-519-adding-a-file-system-path-protocol